storage_controller/src/tenant_shard.rs

```diff
@@ -1351,6 +1352,32 @@ impl TenantShard {
     pub(crate) fn set_preferred_az(&mut self, preferred_az_id: AvailabilityZone) {
         self.preferred_az_id = Some(preferred_az_id);
     }
+
+    pub(crate) fn attached_locations(&self) -> Vec<(NodeId, Generation)> {
+        self.observed
+            .locations
+            .iter()
+            .filter_map(|(node_id, observed)| {
+                let conf = observed.conf.as_ref()?;
+                let is_attached = matches!(
+                    conf.mode,
+                    LocationConfigMode::AttachedMulti
+                        | LocationConfigMode::AttachedSingle
+                        | LocationConfigMode::AttachedStale
+                );
+                let has_generation = conf.generation.is_some();
+                if is_attached && has_generation {
+                    Some((*node_id, conf.generation.unwrap()))
+                } else {
+                    None
+                }
```

Comment:
```suggestion
                use LocationConfiMode::{AttachedMulti,AttachedSingle,AttachedStale};
                
                let conf = observed.conf.as_ref()?;
                
                match (conf.generation, conf.mode) {
                  (Some(gen), AttachedMulti | AttachedSingle | AttachedStale) => Some((*node_id, gen)),
                  _ => None
                }
```

Or

```suggestion
                use LocationConfiMode::{AttachedMulti,AttachedSingle,AttachedStale};
                
                let conf = observed.conf.as_ref()?;
                
                conf.generation
                  .map(|g| (g, conf.mode))
                  .filter(|(_, m)| matches!(m, AttachedMulti | AttachedSingle | AttachedStale))
                  .map(|(g, _)| (*node_id, g))
```

But do we really have support for legacy tenants that do not have generation numbers in storage controller? Seems extra, and I cannot really see for what else could that state exist for.

Reply:
> But do we really have support for legacy tenants that do not have generation numbers in storage controller? 

It's not about legacy tenants, but about detached tenants. Such tenant shards don't have a generation in the database IIUC and we wish to ignore them here.

Reply:
> Such tenant shards don't have a generation in the database IIUC and we wish to ignore them here.

They must have a generation in the database because when they get attached again they'll "continue" from where they were last time in the timeline of generations...

The field has no comments.. Maybe we should come up with one? I think it would sound plausible that one would have no generation when **detaching**, though I am unsure what is the benefit.

Reply:
Oh, I think you were suggesting to skip the `Generation` type here. I guess I could, but it's not actually too unwieldy at the call site and it's more readable than a `u32`.

Reply:
https://github.com/neondatabase/neon/pull/9237/commits/985de0c4b4ea0f3dc5c576770b9752f0e2a85686

Reply:
The `Generation` vs. `u32` was not on my mind at all actually :)

But yeah, thanks for accepting either of suggestions. The `None` case is a bit weird... I guess we found that it's only for the "storcon manages a secondary for cplane attached location" case, but it took some looking.